src/duckling/canvas/tools/entity-move-tool.ts

```diff
@@ -20,51 +21,65 @@ import {drawRectangle} from '../drawing';
 
 import {BaseTool, CanvasMouseEvent, CanvasKeyEvent} from './base-tool';
 import {minCornerSnapDistance} from './_grid-snap';
+import {SnapToGridService} from './grid-snap.service';
 
 @Injectable()
 export class EntityMoveTool extends BaseTool {
-    private _selection : EntityKey;
+    private _selectedEntityKey : EntityKey;
+
     private _mergeKey : any;
-    private _selectOffsetCoords : Vector = {x: 0, y: 0};
+    private _initalEntity : Entity;
+    private _initialMouseLocation : Vector;
 
     constructor(private _entitySelectionService: EntitySelectionService,
                 private _entitySystemService : EntitySystemService,
                 private _entityPositionService : EntityPositionService,
                 private _selectionService : SelectionService,
+                private _snapToGridService : SnapToGridService,
                 private _entityBoxService : EntityBoxService,
                 private _projectService : ProjectService) {
         super();
     }
 
     getDisplayObject(canvasZoom : number) : DisplayObject {
-        return this._drawSelectionBox(this._getSelectedEntityBox(), canvasZoom);
+        let container = new PIXI.Container();
+        let entityBox = this._getSelectedEntityBox();
+        if (entityBox) {
+            container.addChild(this._drawSelectionBox(entityBox, canvasZoom));
+        }
+        return container;
     }
 
     onStageDown(event : CanvasMouseEvent) {
-        this._selection = this._entitySelectionService.getEntityKey(event.stageCoords);
-        this._selectOffsetCoords = {x: 0, y: 0};
-        if (this._selection) {
-            let entity = this._entitySystemService.getEntity(this._selection);
-            let position = this._entityPositionService.getPosition(entity);
-            this._selectOffsetCoords = vectorSubtract(position, event.stageCoords);
-        }
+        this._selectedEntityKey = this._entitySelectionService.getEntityKey(event.stageCoords);
         this._mergeKey = newMergeKey();
-        this._selectionService.select(this._selection, this._mergeKey);
+        this._selectionService.select(this._selectedEntityKey, this._mergeKey);
+
+        if (this._selectedEntityKey) {
+            this._initalEntity = this._entitySystemService.getEntity(this._selectedEntityKey);
+            this._initialMouseLocation = event.stageCoords;
+        }
+
+        return !!this._selectedEntityKey;
     }
 
     onStageMove(event : CanvasMouseEvent) {
-        if (this._selection) {
-            let destination = vectorRound(vectorAdd(event.stageCoords, this._selectOffsetCoords));
-            let gridSize = this._projectService.project.getValue().currentMap.gridSize;
-            if (!this._isSnapToGrid(event)) {
-                destination = vectorAdd(destination, minCornerSnapDistance(
-                    destination, 
-                    this._getSelectedEntityBox(), 
-                    {x: gridSize, y: gridSize}));
+        if (this._selectedEntityKey) {
```

Comment:
Should be an early quit:

`if (!this._selectedEntityKey) {
   return;
}`

to avoid the entire function being indented